src/charm.py

```diff
@@ -90,6 +91,10 @@ def _write_cloud_config(self, data: dict[str, str]) -> None:
         since v2 was removed a long time ago (Queens release)
         https://docs.openstack.org/keystone/latest/contributor/http-api.html
         """
+        os.makedirs(os.path.dirname(OS_CLIENT_CONFIG), exist_ok=True)
```

Comment:
non-blocker:

I prefer to use pathlib (modern version of os) when possible. This case could do:
```python
OS_CLIENT_CONFIG_CACERT = Path(f"/var/snap/{SNAP_NAME}/common/cacert.pem")

OS_CLIENT_CONFIG_CACERT.parent.mkdir(parents=True, exist_ok=True)
OS_CLIENT_CONFIG_CACERT.write_text(self.config["ssl_ca"])
```

Reply:
Done